src/terminal/input/terminalInput.cpp

```diff
@@ -554,6 +554,14 @@ bool TerminalInput::HandleKey(const IInputEvent* const pInEvent)
     {
         const auto seq = _GenerateWin32KeySequence(keyEvent);
         _SendInputSequence(seq);
+
+        // GH#1304 - On <Escape> or <Return> key input, switch IME state to alphanumeic mode.
+        auto key = keyEvent.GetCharData();
+        if (keyEvent.IsKeyDown() && (key == VK_RETURN || key == VK_ESCAPE) &&!keyEvent.IsModifierPressed())
+        {
+            CheckAlphaMode(key);

```

Comment:
I don't believe this is the right spot to put this code. Windows Terminal's IME is implemented here: https://github.com/microsoft/terminal/blob/191eb00f43cce6f0a7c35404937ab1a870aef775/src/cascadia/TerminalControl/TSFInputControl.cpp

It emits an "CompositionCompleted" event that's caught here: https://github.com/microsoft/terminal/blob/191eb00f43cce6f0a7c35404937ab1a870aef775/src/cascadia/TerminalControl/TermControl.cpp#L2577-L2585

In that function you could consider doing any manipulation that you consider necessary (I've never used an IME day to day personally, nor has anyone else on the team, hence the lack of personal experience in how it should behave).

At the start of the TSF control there's also the spot where we set the input mode: https://github.com/microsoft/terminal/blob/191eb00f43cce6f0a7c35404937ab1a870aef775/src/cascadia/TerminalControl/TSFInputControl.cpp#L28-L31

Reply:
Thanks for kind advice. Can you help me on this? I am trying to to use `editContext.InputScope(CoreTextInputScope::AlphanumericHalfWidth);` instead of IMM32 call. 

But there is problem. TSFIntputControl can not see the  \<Return>, \<Escape> keypress.

 The `TSFIntputControl` events to handle composition (like `_compositionCompletedHandler, _textUpdatingHandler, _SendAndClearText`) are fired on CJK input key press, but not fired on alphanumeric input or \<Return>, \<Escape>, I guess. So `TSFIntputControl` can not handle the  \<Return>, \<Escape> keypress. Calling `editContext.InputScope(CoreTextInputScope::AlphanumericHalfWidth)` on every `_compositionCompletedHandler` event is not correct behavior. 

So  \<return>, \<Escape> keypress should be handled by  `ControlCore::SendCharEvent()` or `TermControl::_KeyDownHandle()`  Then to call `editContext.InputScope(CoreTextInputScope::AlphanumericHalfWidth)` on seeing <Return>, \<Escape>,  keypress,  ControlCore` or `TermControl` have to access to`_editConext` or `TSFInputControl` Instance handle. These handles are not available to `ControlCore` or `TermControl`.

Should I register (save) the current _editConext` or `TSFInputControl` instance handle to ControlCore class instance as new member variable and retrieve in `ControlCore` or `TermControl`? I am not sure that is correct way to call `editContext.InputScope(CoreTextInputScope::AlphanumericHalfWidth)`.

BTW, which is better place to handle  \<Return>, \<Escape> keypress ? `ControlCore::SendCharEvent()` or `TermControl::_KeyDownHandle()` ? I guess that TermControl::_KeyDownHandle() seems to be better.

Reply:
> First of all, The `EditContext.InputScope(CoreTextInputScope::AlphanumericHalfWidth)` itself is not designed or intended to end/exit CJK composition conversation. It is deigned to switching between full-width alphanumeric input and half-width alphanumeric input in same composition session. So it does not exit or end CJK composition conversation session.

To be entirely honest, I find the `Windows.UI.Text.Core` API (`CoreTextInput` etc.) deeply confusing and kind of buggy when it comes to Korean and Vietnamese, so I might be of little help. In fact I didn't even know that this API can't technically be used for this.

I have access to its implementation, but that one is just as confusing. The old TSF API is pretty simple to understand and I _think_ `SetInputScope(hwnd, IS_ALPHANUMERIC_HALFWIDTH)` does effectively the same thing, but it probably wouldn't work here, because to me it seems like the WinRT API isn't just a mere wrapper. It holds its own state internally and using the HWND based API will probably not update the WinRT state... I think?
We have a TSF implementation using the old API under `/src/tsf` and while that implementation is pretty verbose it's actually mostly just whitespace and comments. It ends up being pretty easy to debug and seems rather robust. I'm saying all this because when I said:
> would it make sense to continue using the CoreTextInput APIs?

I didn't want to imply that you _have_ to use it. It would be better for consistency, but if you find the `Windows.UI.Text.Core` API entirely unsuitable, I would not mind approving this PR if it used a different API. It's just preferable for consistency IMO.

---

> But there is problem. TSFIntputControl can not handle the `<Return>`, `<Escape>` keypress.

You could add a method to `TSFIntputControl` like "ResetInputScope" or something (I'm not sure what name would be best) and call that function whenever you encounter a Return/Escape key press in `TermControl`.

---

> BTW, which is better place to handle <Return>, <Escape> keypress ? ControlCore::SendCharEvent() or TermControl::_KeyDownHandle() ? I guess that TermControl::_KeyDownHandle() seems to be better.

Yes, I think so too. If you call `TSFInputControl()` inside the `TermControl` class you get a pointer to the `TSFInputControl` instance, which would then allow you to call any methods on it you'd like, like "ResetInputScope" for instance.

Reply:
Normally, CJK input is handled largely by operation system and IME, not by individual application, on OS platform like Windows, Linux, Android, iOS/MacOS. It is quite rare case that application handles CJK input directly, except few big-name commercial application like MS Word. I am impressed with handling of CJK input in this MS Open Terminal. 

But, I haven't seen any cross-platform application that handles CJK input by itself, without relying one IME. CJK developers and users do not expect such input functionality. Many Linux applications ported to Win32 platform like X-window/GTK application or text editors or terminal emulator like MinTTY work that way. I personally have added same feature like this PR to MinTTY and iPuTTY using above IMM routines.

So using IMM routines to add such feature is not committing unholy/deadly sin or blasphemy, I guess. IMHO, it is well-acceptable compromise on cross-platform portability vs. usability.